storage/transfermanager/integration_test.go

```diff
@@ -456,6 +457,288 @@ func TestIntegration_DownloadDirectoryError(t *testing.T) {
+				t.Errorf("object(%q) at filepath(%q): content crc32c does not match; got: %v, expected: %v", got.Object, path, gotCRC, wantCRC)
+			}
+		}
+	})
+}
+
+func TestIntegration_DownloadDirectoryWithSkippedFilesAsync(t *testing.T) {
+	multiTransportTest(context.Background(), t, func(t *testing.T, ctx context.Context, c *storage.Client, tb downloadTestBucket) {
+		localDir := t.TempDir()
+		illegalObjs := []string{
+			"../objA",
+			"../objB",
+		}
+		b := c.Bucket(tb.bucket)
+		for _, obj := range illegalObjs {
+			size := randomInt64(minObjectSize, maxObjectSize)
+			_, err := generateFileInGCS(ctx, b.Object(obj), size)
+			if err != nil {
+				t.Errorf("could not create file in GCS: %v for obj %s", err, obj)
+			}
+		}
+		t.Cleanup(func() {
+			for _, obj := range illegalObjs {
+				if err := b.Object(obj).Delete(ctx); err != nil {
+					t.Errorf("failed to delete object in cleanup for %s: %v", obj, err)
+				}
+			}
+		})
+		d, err := NewDownloader(c, WithWorkers(2), WithPartSize(maxObjectSize/2), WithCallbacks())
+		if err != nil {
+			t.Fatalf("NewDownloader: %v", err)
+		}
+		objectDownloaded := make(chan bool)
+		objectSkipped := make(chan bool)
+		done := make(chan bool)
+
+		trackObjectsDownloaded := func(objectsDownloadedCount, objectSkippedCount *int) {
+			for {
+				select {
+				case <-done:
+					return
+				case <-objectDownloaded:
+					*objectsDownloadedCount++
+				case <-objectSkipped:
+					*objectSkippedCount++
+				}
+			}
+		}
+
+		objectsDownloadedCount := 0
+		objectSkippedCount := 0
+		go trackObjectsDownloaded(&objectsDownloadedCount, &objectSkippedCount)
+
+		if err := d.DownloadDirectory(ctx, &DownloadDirectoryInput{
+			Bucket:         tb.bucket,
+			LocalDirectory: localDir,
+			OnObjectDownload: func(got *DownloadOutput) {
+				if got.Err != nil {
+					objectSkipped <- true
+				} else {
+					objectDownloaded <- true
+				}
+
+				if slices.Contains(illegalObjs, got.Object) {
+					if got.Err == nil {
+						t.Errorf("Expected error but got nil for object %v: %v", got.Object, got.Err)
+					}
+					return
+				}
+
+				if got.Err != nil {
+					t.Errorf("result.Err: %v", got.Err)
+				}
+
+				if got, want := got.Attrs.Size, tb.objectSizes[got.Object]; want != got {
+					t.Errorf("expected object size %d, got %d", want, got)
+				}
+
+				path := filepath.Join(localDir, got.Object)
+
+				f, err := os.Open(path)
+				if err != nil {
+					t.Errorf("os.Open(%q): %v", path, err)
+				}
+				defer f.Close()
+
+				b := bytes.NewBuffer(make([]byte, 0, got.Attrs.Size))
+				if _, err := io.Copy(b, f); err != nil {
+					t.Errorf("io.Copy: %v", err)
+				}
+
+				if wantCRC, gotCRC := tb.contentHashes[got.Object], crc32c(b.Bytes()); gotCRC != wantCRC {
+					t.Errorf("object(%q) at filepath(%q): content crc32c does not match; got: %v, expected: %v", got.Object, path, gotCRC, wantCRC)
+				}
+				got.Object = "modifying this shouldn't be a problem"
+			},
+			Callback: func(outs []DownloadOutput) {
```

Comment:
For the async case maybe let's just test that the callback shows the errors we expect for the illegal files and that the other objects got downloaded successfully? IMO this test is duplicating work done elsewhere, let's pare it back somewhat.